### PR TITLE
PARQUET-516: Better error handling in LocalFileSource

### PR DESCRIPTION
**Pull request by asandryh**
Also, a fix for PARQUET-537: Ensure LocalFileSource is closed properly.

### CODE REVIEW

File: src/parquet/file/reader-internal.cc [hunk 1]
```diff
@@ -217,6 +217,10 @@ std::unique_ptr<ParquetFileReader::Contents> SerializedFile::Open(
   return result;
 }
 
+SerializedFile::~SerializedFile() {
+  Close();
+}
+
```
Comment (asandryh) [inline src/parquet/file/reader-internal.cc]: This addresses PARQUET-537

File: src/parquet/file/reader.cc [hunk 2]
```diff
@@ -74,13 +74,16 @@ RowGroupStatistics RowGroupReader::GetColumnStats(int i) const {
 // ParquetFileReader public API
 
 ParquetFileReader::ParquetFileReader() : schema_(nullptr) {}
-ParquetFileReader::~ParquetFileReader() {}
+ParquetFileReader::~ParquetFileReader() {
+  Close();
+}
```
Comment (asandryh) [inline src/parquet/file/reader.cc]: This addresses PARQUET-537

**Comment by asandryh** [comment]
@wesm Can you look at this PR and let me know whether this is what you had in mind for PARQUET-516? Thanks.

**Comment by wesm** [comment]
Yes, but we need to write unit tests to verify the correct functioning of these classes and their methods.

**Comment by asandryh** [comment]
@wesm Right, will do. Just wanted to check first whether you had anything else in mind.

**Comment by wesm** [comment]
In the tests the main thing will be failing in the expected ways (e.g. exceptions raised instead of core dump / segfault later).

**Comment by wesm** [comment]
I'm working on a patch for PARQUET-520 and will add unit tests for the rest of input.h/cc.

**Comment by asandryh** [comment]
@wesm I've updated this PR with unit tests. Please take a look.

File: src/parquet/util/input.cc [hunk 3]
```diff
@@ -54,23 +58,47 @@ void LocalFileSource::Close() {
...
   return size_;
 }
 
+void LocalFileSource::SeekFile(int64_t pos, int origin) const {
+  if (origin == SEEK_SET && pos >= size_) {
+    std::stringstream ss;
+    ss << "Cannot seek to position " << pos << " in file [" << path_
+       << "] of size " << size_;
+    throw ParquetException(ss.str());
+  }
```
Comment (asandryh) [inline src/parquet/util/input.cc]: Some implementations of fseek() will not produce an error if you try to seek past the end of the file. Hence, we should track this explicitly.

**Comment by wesm** [comment]
Argh, we have crossed streams.

**Comment by wesm** [comment]
Can you create a separate patch for PARQUET-537? I will incorporate the additional unit tests here into my patch-in-progress (which also includes a MemoryMapSource).

**Comment by wesm** [comment]
Please see #66 and let me know if I missed any test cases here

**Pull request closed without merging**